In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn import linear_model
import matplotlib.pyplot as plt
import statsmodels.api as sm
from sqlalchemy import create_engine

# Display preferences.
%matplotlib inline
pd.options.display.float_format = '{:.3f}'.format
pd.set_option('display.max_columns', 100)

import warnings
warnings.filterwarnings('ignore')

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))
df = pd.read_sql_query('select * from weatherinszeged',con=engine)

engine.dispose()

df.head(20)

date        summary preciptype  temperature  \
0  2006-03-31 22:00:00+00:00  Partly Cloudy       rain        9.472   
1  2006-03-31 23:00:00+00:00  Partly Cloudy       rain        9.356   
2  2006-04-01 00:00:00+00:00  Mostly Cloudy       rain        9.378   
3  2006-04-01 01:00:00+00:00  Partly Cloudy       rain        8.289   
4  2006-04-01 02:00:00+00:00  Mostly Cloudy       rain        8.756   
5  2006-04-01 03:00:00+00:00  Partly Cloudy       rain        9.222   
6  2006-04-01 04:00:00+00:00  Partly Cloudy       rain        7.733   
7  2006-04-01 05:00:00+00:00  Partly Cloudy       rain        8.772   
8  2006-04-01 06:00:00+00:00  Partly Cloudy       rain       10.822   
9  2006-04-01 07:00:00+00:00  Partly Cloudy       rain       13.772   
10 2006-04-01 08:00:00+00:00  Partly Cloudy       rain       16.017   
11 2006-04-01 09:00:00+00:00  Partly Cloudy       rain       17.144   
12 2006-04-01 10:00:00+00:00  Partly Cloudy       rain       17.800   
13 2006-04-01 11:00:00+00:00  Partly Cloudy       rain       17.333   
14 2006-04-01 12:00:00+00:00  Partly Cloudy       rain       18.878   
15 2006-04-01 13:00:00+00:00  Partly Cloudy       rain       18.911   
16 2006-04-01 14:00:00+00:00  Partly Cloudy       rain       15.389   
17 2006-04-01 15:00:00+00:00  Mostly Cloudy       rain       15.550   
18 2006-04-01 16:00:00+00:00  Mostly Cloudy       rain       14.256   
19 2006-04-01 17:00:00+00:00  Mostly Cloudy       rain       13.144   

    apparenttemperature  humidity  windspeed  windbearing  visibility  \
0                 7.389     0.890     14.120      251.000      15.826   
1                 7.228     0.860     14.265      259.000      15.826   
2                 9.378     0.890      3.928      204.000      14.957   
3                 5.944     0.830     14.104      269.000      15.826   
4                 6.978     0.830     11.045      259.000      15.826   
5                 7.111     0.850     13.959      258.000      14.957   
6                 5.522     0.950     12.365      259.000       9.982   
7                 6.528     0.890     14.152      260.000       9.982   
8                10.822     0.820     11.318      259.000       9.982   
9                13.772     0.720     12.526      279.000       9.982   
10               16.017     0.670     17.565      290.000      11.206   
11               17.144     0.540     19.787      316.000      11.447   
12               17.800     0.550     21.944      281.000      11.270   
13               17.333     0.510     20.689      289.000      11.270   
14               18.878     0.470     15.376      262.000      11.447   
15               18.911     0.460     10.401      288.000      11.270   
16               15.389     0.600     14.409      251.000      11.270   
17               15.550     0.630     11.157      230.000      11.447   
18               14.256     0.690      8.517      163.000      11.206   
19               13.144     0.700      7.631      139.000      11.206   

    loudcover  pressure                       dailysummary  
0       0.000  1015.130  Partly cloudy throughout the day.  
1       0.000  1015.630  Partly cloudy throughout the day.  
2       0.000  1015.940  Partly cloudy throughout the day.  
3       0.000  1016.410  Partly cloudy throughout the day.  
4       0.000  1016.510  Partly cloudy throughout the day.  
5       0.000  1016.660  Partly cloudy throughout the day.  
6       0.000  1016.720  Partly cloudy throughout the day.  
7       0.000  1016.840  Partly cloudy throughout the day.  
8       0.000  1017.370  Partly cloudy throughout the day.  
9       0.000  1017.220  Partly cloudy throughout the day.  
10      0.000  1017.420  Partly cloudy throughout the day.  
11      0.000  1017.740  Partly cloudy throughout the day.  
12      0.000  1017.590  Partly cloudy throughout the day.  
13      0.000  1017.480  Partly cloudy throughout the day.  
14      0.000  1017.170  Partly cloudy throughout the day.  
15      0.000  1016.47

In [2]:
df['tempdifference'] = df['apparenttemperature'] - df['temperature']

# Y is the target variable
Y = df['tempdifference']
# X is the feature set
X = df[['humidity','windspeed']]

X = sm.add_constant(X)

# We fit an OLS model using statsmodels
results = sm.OLS(Y, X).fit()

# We print the summary results
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:         tempdifference   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Thu, 28 Nov 2019   Prob (F-statistic):               0.00
Time:                        14:52:37   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.0

The R-squared and adjusted R-squared values are both 0.288. These seem low

In [3]:
df['wind_humidity'] =  df['humidity']* df['windspeed']

# Y is the target variable
Y = df['tempdifference']
# X is the feature set
X = df[['humidity','windspeed','wind_humidity']]

X = sm.add_constant(X)

# We fit an OLS model using statsmodels
results = sm.OLS(Y, X).fit()

# We print the summary results
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:         tempdifference   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Thu, 28 Nov 2019   Prob (F-statistic):               0.00
Time:                        14:57:25   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const             0.0839      0.033      2.511

Now both R-values are 0.341, which is an improvement.

In [4]:
# Y is the target variable
Y = df['tempdifference']
# X is the feature set
X = df[['humidity','windspeed','visibility']]

X = sm.add_constant(X)

# We fit an OLS model using statsmodels
results = sm.OLS(Y, X).fit()

# We print the summary results
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:         tempdifference   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Thu, 28 Nov 2019   Prob (F-statistic):               0.00
Time:                        15:03:50   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.5756      0.028     56.605      0.0

This is an improvement over the first model in terms of adjusted R-squared. However, the second model is still better in terms of adjusted R-squared.

Considering AIC and BIC, the second model also has the lowest for both.